File: agentic_os/mission/lifecycle.py
"""Mission lifecycle contributors — host-agnostic hooks the runtime dispatches at key phases.

The contract (from xai-grok-build's agent lifecycle): a contributor **receives data-only inputs at
dispatch time, acts only through capabilities injected at install time, and never owns loop
control**. So a contributor cannot alter the plan, block a node, or steer the scheduler — it observes
(and may act through an injected capability, e.g. append evidence, emit a metric, wake the monitor),
and the Mission Runtime keeps the loop. This is the same invariant as the Context Runtime resolve
door: orchestration decides; extensions state needs and act through granted capabilities.

The registry is optional and no-op when empty — existing runtimes are unaffected. A contributor that
raises is swallowed (an extension must never break orchestration).
"""
from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Callable


# ── data-only lifecycle events (no methods, no runtime handles) ──
@dataclass(frozen=True)
class MissionStarted:
    mission_id: str
    goal: str
    template: str | None = None


@dataclass(frozen=True)
class NodeCompleted:
    mission_id: str
    node_id: str
    capability: str
    status: str


@dataclass(frozen=True)
class GateReached:
    mission_id: str
    node_id: str
    capability: str


@dataclass(frozen=True)
class MissionFinished:
    mission_id: str
    state: str          # succeeded | failed


@dataclass(frozen=True)
class SessionIdle:
    reason: str = "idle"


# event type → the contributor method that handles it
_HANDLER = {
    MissionStarted: "on_mission_started",
    NodeCompleted: "on_node_completed",
    GateReached: "on_gate_reached",
    MissionFinished: "on_mission_finished",
    SessionIdle: "on_session_idle",
}


class LifecycleContributor:
    """Base class — override the hooks you care about. Each receives ``(event, capabilities)`` where
    ``capabilities`` is the dict of injected data-services the contributor may act through. Return
    value is ignored (no loop control)."""

    def on_mission_started(self, event: MissionStarted, capabilities: dict[str, Any]) -> None: ...
    def on_node_completed(self, event: NodeCompleted, capabilities: dict[str, Any]) -> None: ...
    def on_gate_reached(self, event: GateReached, capabilities: dict[str, Any]) -> None: ...
    def on_mission_finished(self, event: MissionFinished, capabilities: dict[str, Any]) -> None: ...
    def on_session_idle(self, event: SessionIdle, capabilities: dict[str, Any]) -> None: ...


class LifecycleRegistry:
    """Holds contributors + the injected capabilities, and dispatches events to them. No-op when
    empty. Capabilities are injected once at install time; events carry only data at dispatch."""

    def __init__(self, capabilities: dict[str, Any] | None = None):
        self._contributors: list[Any] = []
        self.capabilities: dict[str, Any] = capabilities or {}

    def install(self, contributor: Any) -> "LifecycleRegistry":
        self._contributors.append(contributor)
        return self

    def inject(self, name: str, capability: Any) -> "LifecycleRegistry":
        self.capabilities[name] = capability
        return self

    def dispatch(self, event: Any) -> None:
        method = _HANDLER.get(type(event))
        if method is None:
            return
        for c in self._contributors:
            fn: Callable | None = getattr(c, method, None)
            if fn is None:
                continue
            try:
                fn(event, self.capabilities)          # data-only in; return ignored (no loop control)
            except Exception:  # noqa: BLE001 — an extension must never break the mission loop
                pass

File: go/mission/lifecycle.go
package mission

// Mission lifecycle contributors (Go port of lifecycle.py). A contributor receives data-only inputs
// at dispatch time, acts only through capabilities injected at install time, and never owns loop
// control. Optional + no-op when empty; a contributor that panics is recovered (an extension must
// never break the mission loop).

// ── data-only lifecycle events ──
type MissionStarted struct {
	MissionID string
	Goal      string
	Template  string
}
type NodeCompleted struct {
	MissionID  string
	NodeID     string
	Capability string
	Status     string
}
type GateReached struct {
	MissionID  string
	NodeID     string
	Capability string
}
type MissionFinished struct {
	MissionID string
	State     string // succeeded | failed
}
type SessionIdle struct {
	Reason string
}

// LifecycleContributor — set the hooks you care about; nil hooks are simply not called. Each
// receives the event + the injected capabilities dict (the data-services it may act through).
type LifecycleContributor struct {
	OnMissionStarted  func(MissionStarted, map[string]any)
	OnNodeCompleted   func(NodeCompleted, map[string]any)
	OnGateReached     func(GateReached, map[string]any)
	OnMissionFinished func(MissionFinished, map[string]any)
	OnSessionIdle     func(SessionIdle, map[string]any)
}

// LifecycleRegistry holds contributors + injected capabilities and dispatches events. No-op when empty.
type LifecycleRegistry struct {
	contributors []LifecycleContributor
	Capabilities map[string]any
}

func NewLifecycleRegistry(capabilities map[string]any) *LifecycleRegistry {
	if capabilities == nil {
		capabilities = map[string]any{}
	}
	return &LifecycleRegistry{Capabilities: capabilities}
}

func (r *LifecycleRegistry) Install(c LifecycleContributor) *LifecycleRegistry {
	r.contributors = append(r.contributors, c)
	return r
}

func (r *LifecycleRegistry) Inject(name string, capability any) *LifecycleRegistry {
	r.Capabilities[name] = capability
	return r
}

// Dispatch delivers a data-only event to every contributor's matching hook (return ignored — no
// loop control). A panicking contributor is recovered so orchestration is never broken.
func (r *LifecycleRegistry) Dispatch(event any) {
	for _, c := range r.contributors {
		func(c LifecycleContributor) {
			defer func() { _ = recover() }()
			switch e := event.(type) {
			case MissionStarted:
				if c.OnMissionStarted != nil {
					c.OnMissionStarted(e, r.Capabilities)
				}
			case NodeCompleted:
				if c.OnNodeCompleted != nil {
					c.OnNodeCompleted(e, r.Capabilities)
				}
			case GateReached:
				if c.OnGateReached != nil {
					c.OnGateReached(e, r.Capabilities)
				}
			case MissionFinished:
				if c.OnMissionFinished != nil {
					c.OnMissionFinished(e, r.Capabilities)
				}
			case SessionIdle:
				if c.OnSessionIdle != nil {
					c.OnSessionIdle(e, r.Capabilities)
				}
			}
		}(c)
	}
}
